src/backends/cuda.jl

```diff
@@ -160,9 +160,36 @@ function launch_config(kernel::Kernel{CUDADevice}, ndrange, workgroupsize)
     return ndrange, workgroupsize, iterspace, dynamic
 end
 
+function threads_to_workgroupsize(threads, ndrange)
+    workgroupsize = ones(Int, length(ndrange))
+    workgroupsize[1] = threads
+    i = 1
+    for outer i in 1:length(ndrange)-1
+        workgroupsize[i] > ndrange[i] || break
+        workgroupsize[i+1] = div(threads, prod(ndrange[1:i]))
+        workgroupsize[i] = ndrange[i]
+    end
+    workgroupsize[i] = min(workgroupsize[i], ndrange[i])
+    return Tuple(workgroupsize)
+end
+
 function (obj::Kernel{CUDADevice})(args...; ndrange=nothing, dependencies=nothing, workgroupsize=nothing, progress=yield)
 
-    ndrange, workgroupsize, iterspace, dynamic = launch_config(obj, ndrange, workgroupsize)
+    ndrange, _workgroupsize, iterspace, dynamic = launch_config(obj, ndrange, workgroupsize)
+    kernel = nothing
+
+    if KernelAbstractions.workgroupsize(obj) <: DynamicSize && workgroupsize === nothing
+        ndrange, _, iterspace, _ = launch_config(obj, ndrange, ndrange)
```

Comment:
Why can't we re-use the result of `launch_config` three lines above?

Reply:
The problem is that the type of `workgroupsize` is statically embedded in the cassette pass, so to be able to use the already compiled `cufunction`, the preliminary `workgroupsize` we pass here has the same number of dimensions than our eventual optimized `workgroupsize`. Since the actual value doesn't really matter, I just used `ndrange` here, and then `threads_to_workgroupsize` always returns a tuple of the same length as `ndrange`. My current implementation is structured a bit weird, I will try to make this a little more obvious.